### PR TITLE
Update to EVM with block mapping

### PR DESCRIPTION
**Pull request by sideninja**
Closes: #???

## Description
This PR updates flow-go to latest version that includes block mapping and makes corresponding changes to the logic and ensures backward compatibility.
______

For contributor use:

- [ ] Targeted PR against `master` branch
- [ ] Linked to Github issue with discussion and accepted design OR link to spec that describes this work.
- [ ] Code follows the [standards mentioned here](https://github.com/onflow/flow-nft/blob/master/CONTRIBUTING.md#styleguides).
- [ ] Updated relevant documentation 
- [ ] Re-reviewed `Files changed` in the Github PR explorer
- [ ] Added appropriate labels 

## Summary by CodeRabbit

- **New Features**
	- Enhanced block handling with a new model structure for improved clarity and functionality.
	- Introduction of new methods for decoding and processing Cadence events related to blocks and transactions.
  
- **Bug Fixes**
	- Refined error handling in event processing and block indexing methods to provide more specific error messages.

- **Refactor**
	- Transitioned from using `types.Block` to `models.Block` across various components for improved data handling and consistency.
  
- **Tests**
	- Expanded test coverage to accommodate changes in block structures and event handling, ensuring robustness in processing.
	- Updated assertions in tests to align with the latest blockchain data and expected outcomes.

- **Chores**
	- Updated dependencies to ensure security and stability.

### CODE REVIEW

File: models/events.go [hunk 1]
```diff
@@ -1,114 +1,142 @@
...
 // isTransactionExecutedEvent checks whether the given event contains transaction executed data.
 func isTransactionExecutedEvent(event cadence.Event) bool {
 	if event.EventType == nil {
 		return false
 	}
-	return strings.Contains(event.EventType.ID(), string(types.EventTypeTransactionExecuted))
+	return strings.Contains(event.EventType.ID(), string(events.EventTypeTransactionExecuted))
 }
 
+// CadenceEvents contains Flow emitted events containing one or zero evm block executed event,
```
Comment (m-Peter) [inline models/events.go:28]:
```suggestion
// CadenceEvents contains Flow emitted events containing exactly one evm block executed event,
```

File: models/events.go [hunk 2]
```diff
@@ -1,114 +1,142 @@
...
+// and multiple or zero evm transaction events.
 type CadenceEvents struct {
-	events flow.BlockEvents
+	events       flow.BlockEvents  // Flow events for a specific flow block
+	block        *Block            // EVM block (at most one per Flow block)
```
Comment (m-Peter) [inline models/events.go:32]:
```suggestion
	block        *Block            // EVM block (exactly one per Flow block)
```
I believe we should have exactly one EVM Block per Flow Block. Maybe it is worth it to remove the pointer? To make sure that it is never `nil`.

File: models/events.go [hunk 3]
```diff
@@ -1,114 +1,142 @@
...
-		if isTransactionExecutedEvent(e.Value) {
-			tx, receipt, err := decodeTransactionEvent(e.Value)
-			if err != nil {
-				return nil, nil, err
+	// decode and cache block and transactions
+	for _, event := range events.Events {
+		val := event.Value
+
+		if isBlockExecutedEvent(val) {
+			if e.block != nil { // safety check, we can only have 1 or 0 evm blocks per one flow block
```
Comment (m-Peter) [inline models/events.go:87]:
```suggestion
			if e.block != nil { // safety check, we can have exactly 1 evm block per one flow block
```

File: models/events.go [hunk 4]
```diff
@@ -1,114 +1,142 @@
...
+			}
+
+			e.transactions = append(e.transactions, tx)
+			e.receipts = append(e.receipts, receipt)
 		}
 	}
 
-	return txs, rcps, nil
+	// safety check, we can't have an empty block with transactions
+	if e.block == nil && len(e.transactions) > 0 {
```
Comment (m-Peter) [inline models/events.go:112]: I believe we should add a check to make sure that after all processing, `e.block != nil`, as we should have exactly 1 block for each Flow block, even if without any transactions.

File: models/events.go [hunk 5]
```diff
@@ -1,114 +1,142 @@
...
 func isTransactionExecutedEvent(event cadence.Event) bool {
 	if event.EventType == nil {
 		return false
 	}
-	return strings.Contains(event.EventType.ID(), string(types.EventTypeTransactionExecuted))
+	return strings.Contains(event.EventType.ID(), string(events.EventTypeTransactionExecuted))
 }
 
+// CadenceEvents contains Flow emitted events containing one or zero evm block executed event,
+// and multiple or zero evm transaction events.
```
Comment (m-Peter) [inline models/events.go:29]:
```suggestion
// and zero or multiple evm transaction events.
```

File: models/events.go [hunk 6]
```diff
@@ -1,114 +1,142 @@
...
+		e.block.TransactionHashes = append(e.block.TransactionHashes, rcp.TxHash)
+		// calculate cumulative gas used up to that point
+		cumulativeGasUsed += rcp.GasUsed
+		rcp.CumulativeGasUsed = cumulativeGasUsed
+		// set the transaction index
+		rcp.TransactionIndex = uint(i)
+		// set calculate block hash
+		rcp.BlockHash = blockHash
+		// dynamically add missing log fields
+		for _, l := range rcp.Logs {
```
Comment (m-Peter) [inline models/events.go:67]:
```suggestion
for _, l := range rcp.Logs {
	l.BlockNumber = rcp.BlockNumber.Uint64()
	l.BlockHash = rcp.BlockHash
	l.TxHash = rcp.TxHash
	l.TxIndex = rcp.TransactionIndex
	l.Index = rcp.TransactionIndex
	l.Removed = false
}
```

The consensus fields of `Log` type are only `Address` / `Topics` / `Data`, which means we have to dynamically add the rest. See: https://github.com/onflow/go-ethereum/blob/master/core/types/log.go#L30-L36

File: models/events.go [hunk 7]
```diff
@@ -1,114 +1,142 @@
...
+// Receipts included in the EVM block, if event doesn't
+// contain EVM transactions the return value is nil.
+func (c *CadenceEvents) Receipts() []*StorageReceipt {
+	return c.receipts
+}
+
+// Empty checks if there is an EVM block included in the events.
+// If there are no evm block or transactions events this is a heartbeat event.
+func (c *CadenceEvents) Empty() bool {
+	return c.block == nil
```
Comment (m-Peter) [inline models/events.go:139]: Theoretically, the system chunk transaction is run for each Flow block formation. Which means for each Flow block, there will be emitted an `EVM.BlockExecuted` event. I don't see how the EVM block will be `nil`, for responses from the Event Streaming API. Basically, we should get a response for every single Flow block. I believe `Empty` here means that the EVM block had no EVM transactions.

File: models/events_test.go [hunk 8]
```diff
@@ -5,82 +5,181 @@ import (
...
+	for i := 0; i < txCount; i++ {
+		var err error
+		txs[i], results[i], txEvents[i], err = newTransaction(uint64(i))
+		require.NoError(t, err)
+		hashes[i] = txs[i].Hash()
+		blockEvents.Events = append(blockEvents.Events, txEvents[i])
+	}
+
+	// generate single block
+	block, blockEvent, err := newBlock(1, hashes)
```
Comment (m-Peter) [inline models/events_test.go:89]:
```suggestion
	block, blockEvent, err := newBlock(1, hashes)
	require.NoError(t, err)
```

File: models/events.go (hunk 1 above)
Comment (sideninja) [reply models/events.go:28]: we will leave an option for Flow block to not contain EVM block as per protocol design, we will now have 1 but in future we might not

Comment (m-Peter) [reply models/events.go:28]: Ah, I see. I wasn't aware of that :+1:

**Review by m-Peter — APPROVED** [review]
LGTM! :tada: :100:

**Review by m-Peter — APPROVED** [review]
LGTM! :tada:

**Pull request merged**